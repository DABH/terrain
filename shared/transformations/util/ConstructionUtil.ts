/*
University of Illinois/NCSA Open Source License 

Copyright (c) 2018 Terrain Data, Inc. and the authors. All rights reserved.

Developed by: Terrain Data, Inc. and
              the individuals who committed the code in this file.
              https://github.com/terraindata/terrain
                  
Permission is hereby granted, free of charge, to any person 
obtaining a copy of this software and associated documentation files 
(the "Software"), to deal with the Software without restriction, 
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

* Redistributions of source code must retain the above copyright notice, 
  this list of conditions and the following disclaimers.

* Redistributions in binary form must reproduce the above copyright 
  notice, this list of conditions and the following disclaimers in the 
  documentation and/or other materials provided with the distribution.

* Neither the names of Terrain Data, Inc., Terrain, nor the names of its 
  contributors may be used to endorse or promote products derived from
  this Software without specific prior written permission.

This license supersedes any copyright notice, license, or related statement
following this comment block.  All files in this repository are provided
under the same license, regardless of whether a corresponding comment block
appears in them.  This license also applies retroactively to any previous
state of the repository, including different branches and commits, which
were made public on or after December 8th, 2018.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
THE SOFTWARE.
*/

// Copyright 2018 Terrain Data, Inc.
// tslint:disable:max-classes-per-file

import * as Immutable from 'immutable';
import * as _ from 'lodash';
const { List, Map } = Immutable;

import * as Utils from 'shared/transformations/util/EngineUtils';

import { FieldTypes } from 'shared/etl/types/ETLTypes';
import { TransformationEngine } from 'shared/transformations/TransformationEngine';
import { KeyPath, WayPoint } from 'shared/util/KeyPath';
import * as yadeep from 'shared/util/yadeep';

import * as TerrainLog from 'loglevel';

import { TypeTracker } from './TypeTracker';

interface FieldNode
{
  path: KeyPath;
  name: WayPoint; // null if root
  type?: FieldTypes;
  arrChild?: FieldNode;
  fields: { [k: string]: FieldNode };
}

export default abstract class ConstructionUtil
{
  // take two engines and return an engine whose fields most closely resembles the result of the merge
  public static mergeJoinEngines(
    leftEngine: TransformationEngine,
    rightEngine: TransformationEngine,
    outputKey: string,
  ): TransformationEngine
  {
    const newEngine = new TransformationEngine();
    leftEngine.getAllFieldIDs().forEach((id) =>
    {
      const keypath = leftEngine.getFieldPath(id);
      const newId = Utils.fields.copyField(leftEngine, id, keypath, undefined, newEngine);
    });
    const fieldPathBase = List([outputKey, -1]);
    const outputFieldId = Utils.fields.addFieldToEngine(newEngine, List([outputKey]), FieldTypes.Array);
    const outputFieldWildcardId = Utils.fields.addFieldToEngine(newEngine, fieldPathBase, FieldTypes.Object);

    rightEngine.getAllFieldIDs().forEach((id) =>
    {
      const newKeyPath = fieldPathBase.concat(rightEngine.getFieldPath(id)).toList();
      const newId = Utils.fields.copyField(rightEngine, id, newKeyPath, undefined, newEngine);
    });
    return newEngine;
  }

  public static createEngineFromDocuments(documents: List<object>, interpretText = false):
    { engine: TransformationEngine, errors: string[] }
  {
    const pathTypes: { [k: string]: TypeTracker } = {};
    const errAccumulator = ConstructionUtil.errorAccumulator();
    documents.forEach((doc, docIndex) =>
    {
      for (const leaf of yadeep.traverse(doc, { primitivesOnly: true, arrayLimit: 20 }))
      {
        const { location, value } = leaf;
        const path = Utils.path.convertIndices(location);
        const hash = Utils.path.hash(path);
        if (pathTypes[hash] === undefined)
        {
          pathTypes[hash] = new TypeTracker(path, errAccumulator.fn, interpretText);
        }
        pathTypes[hash].push(value);
      }
    });

    const tree = ConstructionUtil.buildTreeFromPathTypes(pathTypes, errAccumulator.fn);

    const engine = new TransformationEngine();
    for (const match of yadeep.traverse(tree, { primitivesOnly: true }))
    {
      const { value, location } = match;
      if (location.last() === 'type')
      {
        const kp = yadeep.get(tree, location.set(-1, 'path'));
        const type = value;
        if (kp.size !== 0)
        {
          engine.addField(kp, { etlType: type });
        }
      }
    }

    return {
      engine,
      errors: errAccumulator.errors,
    };
  }

  public static makeEngine(doc: object): TransformationEngine
  {
    return ConstructionUtil.createEngineFromDocuments(List([doc])).engine;
  }

  private static buildTreeFromPathTypes(pathTypes: { [k: string]: TypeTracker }, onConflict: (msg: string) => void): FieldNode
  {
    const tree: FieldNode = {
      path: KeyPath([]),
      name: null,
      fields: {},
    };

    const walk = (kp: KeyPath, desiredType: FieldTypes) =>
    {
      let node: FieldNode = tree;
      for (let i = 0; i < kp.size; i++)
      {
        const waypoint = kp.get(i);
        if (waypoint === -1)
        {
          if (node.type === undefined)
          {
            node.type = FieldTypes.Array;
          }
          if (node.type !== FieldTypes.Array)
          {
            const message = `Encountered a ${node.type} field where an array field was expected`;
            node.type = FieldTypes.String;
            node.arrChild = undefined;
            node.fields = {};
            return message;
          }
          else if (node.arrChild === undefined)
          {
            node.type = FieldTypes.Array;
            node.arrChild = {
              path: node.path.push(-1),
              name: -1,
              fields: {},
            };
            node = node.arrChild;
          }
          else
          {
            node = node.arrChild;
          }
        }
        else
        {
          if (node.type === undefined)
          {
            node.type = FieldTypes.Object;
          }
          if (node.type !== FieldTypes.Object)
          {
            const message = `Encountered a ${node.type} field where an object field was expected`;
            node.type = FieldTypes.String;
            node.arrChild = undefined;
            node.fields = {};
            return message;
          }
          else if (node.fields[waypoint] === undefined)
          {
            node.type = FieldTypes.Object;
            node.fields[waypoint] = {
              path: node.path.push(waypoint),
              name: waypoint,
              fields: {},
            };
            node = node.fields[waypoint];
          }
          else
          {
            node = node.fields[waypoint];
          }
        }
      }
      if (Object.keys(node.fields).length > 0 || node.arrChild !== undefined)
      {
        node.type = FieldTypes.String;
        node.fields = {};
        node.arrChild = undefined;
        return `Field with primitive type cannot be an array or object`;
      }
      else
      {
        node.type = desiredType;
      }
      return true;
    };

    for (const key of Object.keys(pathTypes))
    {
      const kp = Utils.path.unhash(key);
      const errors = walk(kp, pathTypes[key].getType());
      if (errors !== true)
      {
        onConflict(errors);
      }
    }
    return tree;
  }

  private static errorAccumulator(maxErrors: number = 5): { numErrors: number, errors: string[], fn: (err) => void }
  {
    const errors = [];
    let total = 0;
    return {
      errors,
      numErrors: total,
      fn: (err) =>
      {
        total++;
        if (errors.length <= maxErrors)
        {
          errors.push(err);
        }
      },
    };
  }
}
