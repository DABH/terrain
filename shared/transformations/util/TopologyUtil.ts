/*
University of Illinois/NCSA Open Source License 

Copyright (c) 2018 Terrain Data, Inc. and the authors. All rights reserved.

Developed by: Terrain Data, Inc. and
              the individuals who committed the code in this file.
              https://github.com/terraindata/terrain
                  
Permission is hereby granted, free of charge, to any person 
obtaining a copy of this software and associated documentation files 
(the "Software"), to deal with the Software without restriction, 
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

* Redistributions of source code must retain the above copyright notice, 
  this list of conditions and the following disclaimers.

* Redistributions in binary form must reproduce the above copyright 
  notice, this list of conditions and the following disclaimers in the 
  documentation and/or other materials provided with the distribution.

* Neither the names of Terrain Data, Inc., Terrain, nor the names of its 
  contributors may be used to endorse or promote products derived from
  this Software without specific prior written permission.

This license supersedes any copyright notice, license, or related statement
following this comment block.  All files in this repository are provided
under the same license, regardless of whether a corresponding comment block
appears in them.  This license also applies retroactively to any previous
state of the repository, including different branches and commits, which
were made public on or after December 8th, 2018.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
THE SOFTWARE.
*/

// Copyright 2018 Terrain Data, Inc.

import * as _ from 'lodash';
import { KeyPath } from 'shared/util/KeyPath';
import * as yadeep from 'shared/util/yadeep';
import EngineUtil from 'shared/transformations/util/EngineUtil';

export type Relation = 'one' | 'many';

/*
General method for seeing if 2 keypaths are "compatible"
Take both keypaths
Match the fronts to find the longest match
e.g.
[foo, bar, -1, X, Y, Z]
[foo, bar, -1, A, B, C]
would have [foo, bar, -1] as the longest common match.

Now take the remaining part of the keypath for each path.
In this case it would be [X, Y, Z] and [A, B, C]
As long as at least one of them is singular (no wildcards), then the two are compatible.

General rule: Compatible keypaths are anything that can be posed as a one to one, many to one, or a one to many.
*/

export default class TopologyUtil
{
  public static isKeyPathSingular(kp: KeyPath, from = 0): boolean
  {
    for (let i = from; i < kp.size; i++)
    {
      if (kp.get(i) === -1)
      {
        return false;
      }
    }
    return true;
  }

  // returns the first position where the keypaths differ
  public static getDifferingBaseIndex(kp1: KeyPath, kp2: KeyPath): number
  {
    let i;
    const minSize = Math.min(kp1.size, kp2.size);
    for (i = 0; i < minSize; i++)
    {
      if (kp1.get(i) !== kp2.get(i))
      {
        break;
      }
    }
    return i;
  }

  /*
   *  do the keypaths represent one-to-one, one-to-many, many-to-one, or many-to-many relationships?
   */
  public static getRelation(kp1: KeyPath, kp2: KeyPath): [Relation, Relation]
  {
    const baseIndex = TopologyUtil.getDifferingBaseIndex(kp1, kp2);
    const relation1 = TopologyUtil.isKeyPathSingular(kp1, baseIndex) ? 'one' : 'many';
    const relation2 = TopologyUtil.isKeyPathSingular(kp2, baseIndex) ? 'one' : 'many';
    return [relation1, relation2];
  }

  /*
   *  Given two keypaths that are one-to-one related, return a function that
   *  translates singular paths that match kp1 to singular paths that match kp2.
   *  E.G. kp1 is [foo, -1, bar] and kp2 is [foo, -1, baz]
   *  createOneToOneMatcher(kp1, kp2)([foo, 3, bar]) ---> [foo, 3, baz]
   */
  public static createOneToOneMatcher(kp1: KeyPath, kp2: KeyPath): (kp: KeyPath) => KeyPath
  {
    const baseIndex = TopologyUtil.getDifferingBaseIndex(kp1, kp2);
    const wildcards = [];
    for (let i = 0; i < baseIndex; i++)
    {
      if (kp1.get(i) === -1)
      {
        wildcards.push(i);
      }
    }
    if (wildcards.length === 0)
    {
      return (kp) => kp2;
    }
    else
    {
      // in the absence of benchmarking to find out what's most efficient
      // assume most times we only do 1 or 2 sets
      return (kp: KeyPath) => {
        let ret = kp2;
        for (const i of wildcards)
        {
          ret = kp2.set(i, kp.get(i));
        }
        return ret;
      };
    }
  }

  /*
   * referenceKP can contain -1
   * matchKP should not contain -1
   */
  public static createLocalMatcher(referenceKP: KeyPath, matchKP: KeyPath): (newKP: KeyPath) => KeyPath
  {
    if (referenceKP.size !== matchKP.size || referenceKP.size === 0)
    {
      return null;
    }

    const replacements: {
      [k: number]: number;
    } = {};

    let maxIndex = -1;

    for (let i = 0; i < referenceKP.size; i++)
    {
      const searchIndex = referenceKP.get(i);
      const matchIndex = matchKP.get(i);
      if (searchIndex !== matchIndex)
      {
        if (typeof searchIndex !== 'number' || typeof matchIndex !== 'number')
        {
          return null;
        }
        else
        {
          replacements[i] = matchIndex;
          maxIndex = i;
        }
      }
    }

    const baseMatchPath = matchKP.slice(0, maxIndex + 1);

    return (newKP: KeyPath) =>
    {
      if (maxIndex === -1)
      {
        return newKP;
      }
      else
      {
        const toTransplant = newKP.slice(maxIndex + 1);
        return baseMatchPath.concat(toTransplant).toList();
      }

    };
  }

  // returns true if two given keypaths represent fields that are "local" to each other.
  // fields are local if they are unambiguously traversable to each other.
  // e.g. [a] is local to [b] and [c, d]
  // [a, -1, b] would be local to [a, -1, c]
  // [a] would not be local to [c, -1, d]
  public static areFieldsLocal(kp1, kp2): boolean
  {
    const lastIndex1: number = kp1.findLastIndex((value, index) => EngineUtil.isWildcardField(kp1, index));
    const concretePath1 = kp1.slice(0, lastIndex1 + 1);
    const lastIndex2: number = kp2.findLastIndex((value, index) => EngineUtil.isWildcardField(kp2, index));
    const concretePath2 = kp2.slice(0, lastIndex2 + 1);

    if (lastIndex2 !== lastIndex1 || !concretePath1.equals(concretePath2))
    {
      return false;
    }
    else
    {
      // check if fields are wildcards themselves?
      return true;
    }
  }
}
