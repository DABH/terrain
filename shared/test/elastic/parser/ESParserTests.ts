/*
University of Illinois/NCSA Open Source License 

Copyright (c) 2018 Terrain Data, Inc. and the authors. All rights reserved.

Developed by: Terrain Data, Inc. and
              the individuals who committed the code in this file.
              https://github.com/terraindata/terrain
                  
Permission is hereby granted, free of charge, to any person 
obtaining a copy of this software and associated documentation files 
(the "Software"), to deal with the Software without restriction, 
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

* Redistributions of source code must retain the above copyright notice, 
  this list of conditions and the following disclaimers.

* Redistributions in binary form must reproduce the above copyright 
  notice, this list of conditions and the following disclaimers in the 
  documentation and/or other materials provided with the distribution.

* Neither the names of Terrain Data, Inc., Terrain, nor the names of its 
  contributors may be used to endorse or promote products derived from
  this Software without specific prior written permission.

This license supersedes any copyright notice, license, or related statement
following this comment block.  All files in this repository are provided
under the same license, regardless of whether a corresponding comment block
appears in them.  This license also applies retroactively to any previous
state of the repository, including different branches and commits, which
were made public on or after December 8th, 2018.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
THE SOFTWARE.
*/

// Copyright 2017 Terrain Data, Inc.

import * as fs from 'fs';
import * as util from 'util';
import * as winston from 'winston';

import SharedUtil from '../../../../shared/Util';
import ESJSONParser from '../../../database/elastic/parser/ESJSONParser';

function getExpectedFile(): string
{
  return __filename.split('.')[0] + '.expected';
}

let expected;

beforeAll(async (done) =>
{
  // TODO: get rid of this monstrosity once @types/winston is updated.
  (winston as any).level = 'debug';

  const contents: any = await new Promise((resolve, reject) =>
  {
    fs.readFile(getExpectedFile(), SharedUtil.promise.makeCallback(resolve, reject));
  });

  expected = JSON.parse(contents);
  done();
});

function testParse(testString: string,
  expectedValue: any,
  expectedErrors: any[] = [])
{
  winston.info('testing \'' + testString + '\'');
  const parser: ESJSONParser = new ESJSONParser(testString);
  const value = parser.getValue();
  const errors = parser.getErrors();

  // winston.info(util.inspect(parser.getValueInfo()));
  // winston.info(util.inspect(parser.getValueInfos()));
  // winston.info(util.inspect(parser.getTokens()));
  winston.info(util.inspect(errors));

  expect(value).toEqual(expectedValue);

  expect(errors.length).toEqual(expectedErrors.length);
  for (let i = 0; i < errors.length; ++i)
  {
    expect(errors[i]).toMatchObject(expectedErrors[i]);
  }
}

test('parse valid json objects', () =>
{
  Object.getOwnPropertyNames(expected).forEach(
    (testName: string) =>
    {
      const testValue: any = expected[testName];

      // test parsing the value using a few spacing options
      testParse(JSON.stringify(testValue), testValue);
      testParse(JSON.stringify(testValue, null, 1), testValue);
      testParse(JSON.stringify(testValue, null, 2), testValue);
      testParse(JSON.stringify(testValue, null, 3), testValue);
      testParse(JSON.stringify(testValue, null, 4), testValue);
    });
});

test('parse invalid json objects', () =>
{
  testParse('string', null,
    [
      {
        isWarning: false,
        token: { charNumber: 0, row: 0, col: 0 },
      },
    ]);

  testParse('fulse', false,
    [
      {
        isWarning: false,
        token: { charNumber: 0, row: 0, col: 0 },
      },
    ]);

  testParse('falsey', false,
    [
      {
        isWarning: false,
        token: { charNumber: 0, row: 0, col: 0 },
      },
    ]);

  testParse('tru', true,
    [
      {
        isWarning: false,
        token: { charNumber: 0, row: 0, col: 0 },
      },
    ]);

  testParse('trueee', true,
    [
      {
        isWarning: false,
        token: { charNumber: 0, row: 0, col: 0 },
      },
    ]);

  testParse('n', null,
    [
      {
        isWarning: false,
        token: { charNumber: 0, row: 0, col: 0 },
      },
    ]);

  testParse('nil', null,
    [
      {
        isWarning: false,
        token: { charNumber: 0, row: 0, col: 0 },
      },
    ]);

  testParse('nullllleray', null,
    [
      {
        isWarning: false,
        token: { charNumber: 0, row: 0, col: 0 },
      },
    ]);

  testParse('["string]', ['string'],
    [
      {
        isWarning: false,
        token: { charNumber: 1, row: 0, col: 1 },
      },
    ]);

  testParse(`{
  "index": "movies",
  "type": "data",
  "from": 0,
  "size": "10"
}`, JSON.parse(`{"from": 0, "index": "movies", "size": "10", "type": "data"}`), []);

  testParse(`{
  "index" : "movies",
  "type" : "data",
  "size" : 10,
  "from" : 0,
  "body" : {
    "query" : {
      "bool" : {
        "must_not" : [ { "match" : { "title": "blah blah"} } ]
      }
    }
  }
}`,
    {
      body: { query: { bool: { must_not: [{ match: { title: 'blah blah' } }] } } },
      from: 0,
      index: 'movies',
      size: 10,
      type: 'data',
    });
});
