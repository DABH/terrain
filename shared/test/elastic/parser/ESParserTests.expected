{
  "regular string" : "foo bar baz",
  "newline string" : "\\n",
  "multiple newline string 1" : "\\nfoo \\n bar \\n baz\\n",
  "multiple newline string 2" : "f''_'s;ddV@&N*(NV%^78#$^%*VN(%/////jasjoo \\n basdfhjkasfhjkar \\n bazasdjkl;as@#HCD$*(CNH^$djkl",
  "multiple escaped string" : "\"\b\f\n\r\t foo \b\f\"\n\r\t bar \b\f\n\r\t\" baz \t\r\n\"\f\b\"",
  "hex escaped string" : "\u7F1A\uFEED\uF00D\uBACD\uBAD9\u123D\u8FAC\u4ADC\u654A\u321B\uFFFF\u0000",
  "mixed escaping" : "\n\u1234\t\uFEED\basdf\u001Fdnwyuqi\b\\\" dhuFH#*f3DF! \n \r \t \uF00D\"\\ adshjk \n",
  "zero" : "0",
  "one" : "1",
  "integer" : "1234567",
  "negative integer" : "-38231768",
  "decimal" : "0.136431",
  "negative decimal" : "-0.98927689",
  "float" : "2345.823768",
  "negative float" : "-71.376251",
  "exponent 0" : "1e3",
  "exponent 1" : "-1e55",
  "exponent 2" : "-0.23213e19",
  "exponent 3" : "0.1e0",
  "exponent 4" : "-3127839.9031276868E-10",
  "exponent 5" : "9.02314e9",
  "true" : true,
  "false" : false,
  "null" : null,
  "empty object" : {},
  "empty list" : [],
  "basic search" : {
    "index" : "movies",
    "type" : "data",
    "from" : 0,
    "size" : "10"
  },
  "search" : {
    "max_score" : null,
    "hits" : [
      {
        "_index" : "movies",
        "_type" : "data",
        "_id" : "72998",
        "_score" : null,
        "_source" : {
          "overview" : "In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.",
          "votecount" : 8218,
          "posterpath" : "/tcqb9NHdw9SWs2a88KCDD4V8sVR.jpg",
          "runtime" : 162,
          "movieid" : 72998,
          "language" : "en",
          "releasedate" : "2009-12-10T00:00:00.000Z",
          "voteaverage" : 7.099999904632568,
          "title" : "Avatar (2009)",
          "revenue" : 2100000000,
          "backdroppath" : "/5XPPB44RQGfkBrbJxmtdndKz05n.jpg",
          "genres" : "Action|Adventure|Sci-Fi|IMAX",
          "popularity" : 8.273819923400879,
          "@version" : "1",
          "tagline" : "Enter the World of Pandora.",
          "status" : "Released",
          "budget" : 237000000,
          "homepage" : "http://www.avatarmovie.com/"
        },
        "sort" : [
          2100000000,
          72998
        ]
      }
    ]
  },
  "search with space in field" : {
    "index" : "movies",
    "type" : "data stuff",
    "size" : 10,
    "from" : 0
  },
  "search 2" : {
    "index" : "movies",
    "type" : "data",
    "size" : 10,
    "from" : 0,
    "body" : {
      "query" : {
        "bool" : {
          "must_not" : [{"match" : {"title" : "blah blah"}}]
        }
      }
    }
  },
  "indices.getMapping" : {
    "movies" : {
      "mappings" : {
        "data" : {
          "properties" : {
            "@version" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "backdroppath" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "budget" : {
              "type" : "long"
            },
            "genres" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "homepage" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "language" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "movieid" : {
              "type" : "long"
            },
            "overview" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "popularity" : {
              "type" : "float"
            },
            "posterpath" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "releasedate" : {
              "type" : "date"
            },
            "revenue" : {
              "type" : "long"
            },
            "runtime" : {
              "type" : "long"
            },
            "status" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "tagline" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "title" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "voteaverage" : {
              "type" : "float"
            },
            "votecount" : {
              "type" : "long"
            }
          }
        }
      }
    }
  },
  "putScript" : {
    "max_score" : null,
    "hits" : [
      {
        "_index" : "movies",
        "_type" : "data",
        "_id" : "72998",
        "_score" : null,
        "_source" : {
          "overview" : "In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.",
          "votecount" : 8218,
          "posterpath" : "/tcqb9NHdw9SWs2a88KCDD4V8sVR.jpg",
          "runtime" : 162,
          "movieid" : 72998,
          "language" : "en",
          "releasedate" : "2009-12-10T00:00:00.000Z",
          "voteaverage" : 7.099999904632568,
          "title" : "Avatar (2009)",
          "revenue" : 2100000000,
          "backdroppath" : "/5XPPB44RQGfkBrbJxmtdndKz05n.jpg",
          "genres" : "Action|Adventure|Sci-Fi|IMAX",
          "popularity" : 8.273819923400879,
          "@version" : "1",
          "tagline" : "Enter the World of Pandora.",
          "status" : "Released",
          "budget" : 237000000,
          "homepage" : "http://www.avatarmovie.com/"
        },
        "sort" : [
          1863000000
        ]
      }
    ]
  },
  "search-1" : {
    "index" : "movies",
    "type" : "data",
    "from" : 0,
    "size" : 10
  },
  "search-2" : {
    "index" : "movies",
    "type" : "data",
    "body" : {
      "from" : 0,
      "size" : 10
    }
  },
  "search-3" : {
    "index" : "movies",
    "type" : "data",
    "body" : {
      "query" : {
        "type" : {"value" : "data"}
      }
    },
    "sort" : ["_score"]
  },
  "search-4" : {
    "index" : "movies",
    "type" : "data",
    "from" : 0,
    "size" : 1000,
    "body" : {
      "query" : {
        "bool" : {
          "must" : {
            "match" : {
              "genres" : "Drama"
            },
            "range" : {
              "releaseDate" : {
                "gt" : 2000,
                "lt" : 2010
              }
            }
          }
        }
      }
    }
  },
  "search-5" : {
    "index" : "movies",
    "type" : "data",
    "size" : 10,
    "body" : {
      "_source" : [
        "movieid",
        "title",
        "releasedate"
      ]
    }
  },
  "search-6" : {
    "index" : "movies",
    "type" : "data",
    "size" : 10,
    "body" : {
      "query" : {
        "bool" : {
          "must_not" : [
            {
              "match" : {
                "title" : "Toy Story (1995)"
              }
            }
          ]
        }
      }
    }
  },
  "search-7" : {
    "index" : "movies",
    "type" : "data",
    "from" : 0,
    "size" : 100,
    "body" : {
      "query" : {
        "bool" : {
          "must" : {
            "match" : {"genres" : "Action"}
          },
          "must_not" : {
            "match" : {
              "title" : "Toy Story (1995)"
            }
          }
        }
      }
    }
  },
  "search-8" : {
    "index" : "movies",
    "type" : "data",
    "from" : 20,
    "size" : 10,
    "body" : {
      "_source" : [
        "movieid",
        "title",
        "releasedate"
      ],
      "sort" : [
        {
          "movieid" : {
            "order" : "desc"
          }
        },
        {
          "releasedate" : {
            "order" : "asc"
          }
        }
      ],
      "query" : {
        "bool" : {
          "filter" : {
            "bool" : {
              "must" : [
                {
                  "range" : {
                    "releasedate" : {
                      "gte" : "2007-03-24"
                    }
                  }
                },
                {
                  "range" : {
                    "releasedate" : {
                      "lt" : "2017-03-24"
                    }
                  }
                }
              ]
            }
          },
          "must_not" : [
            {
              "match" : {
                "movieid" : 2134
              }
            }
          ]
        }
      }
    }
  },
  "search-9" : {
    "index" : "movies",
    "type" : "data",
    "body" : {
      "query" : {},
      "sort" : [
        {
          "revenue" : "desc"
        },
        {
          "movieid" : "asc"
        }
      ]
    },
    "size" : 20
  },
  "search-10" : {
    "index" : "movies",
    "type" : "data",
    "from" : 0,
    "size" : 5,
    "body" : {
      "query" : {
        "bool" : {
          "must" : [
            {
              "match" : {
                "status" : "Released"
              }
            },
            {
              "match" : {
                "language" : "en"
              }
            }
          ],
          "must_not" : [
            {
              "term" : {
                "budget" : 0
              }
            },
            {
              "term" : {
                "revenue" : 0
              }
            }
          ],
          "minimum_should_match" : 1,
          "should" : [
            {
              "match" : {
                "genres" : "Thriller"
              }
            },
            {
              "match" : {
                "genres" : "Action"
              }
            }
          ]
        }
      },
      "sort" : {
        "_script" : {
          "type" : "number",
          "order" : "desc",
          "script" : {
            "stored" : "terrain_PWLScore",
            "params" : {
              "factors" : [
                {
                  "weight" : 0.1,
                  "ranges" : [
                    -100.0,
                    100.0
                  ],
                  "outputs" : [
                    0.0,
                    1.0
                  ]
                },
                {
                  "weight" : 0.2,
                  "a" : 0.0,
                  "b" : 1.0,
                  "numerators" : [
                    [
                      "popularity",
                      1
                    ]
                  ],
                  "denominators" : [],
                  "ranges" : [
                    0.0,
                    2.5,
                    5.0,
                    10.0,
                    50.0
                  ],
                  "outputs" : [
                    0.0,
                    0.2,
                    0.5,
                    0.8,
                    1.0
                  ]
                },
                {
                  "weight" : 0.2,
                  "a" : 25.0,
                  "b" : 10.0,
                  "numerators" : [
                    [
                      "voteaverage",
                      1
                    ]
                  ],
                  "denominators" : [
                    [
                      "votecount",
                      1
                    ]
                  ],
                  "ranges" : [
                    0.0,
                    4.0,
                    5.0,
                    10.0
                  ],
                  "outputs" : [
                    0.0,
                    0.1,
                    0.5,
                    1.0
                  ]
                },
                {
                  "weight" : 0.5,
                  "a" : 0.0,
                  "b" : 1.0,
                  "numerators" : [
                    [
                      "revenue",
                      1
                    ],
                    [
                      "budget",
                      -1
                    ]
                  ],
                  "denominators" : [
                    [
                      "budget",
                      1
                    ]
                  ],
                  "ranges" : [
                    -10.0,
                    1.0,
                    2.0,
                    10.0
                  ],
                  "outputs" : [
                    0.0,
                    0.1,
                    0.5,
                    1.0
                  ]
                }
              ]
            }
          }
        }
      }
    }
  },
  "search-11" : {
    "index" : "movies",
    "type" : "movies",
    "from" : 0,
    "size" : 1000,
    "body" : {
      "query" : {
        "bool" : {
          "should" : {
            "range" : {
              "budget" : {
                "gte" : 100000000
              }
            }
          }
        }
      },
      "sort" : {
        "_script" : {
          "type" : "number",
          "order" : "desc",
          "script" : {
            "stored" : "terrain_PWLScore",
            "params" : {
              "factors" : [
                {
                  "weight" : 1,
                  "a" : 0,
                  "b" : 1,
                  "numerators" : [
                    [
                      "budget",
                      1
                    ]
                  ],
                  "denominators" : [
                  ],
                  "ranges" : [
                    2,
                    15
                  ],
                  "outputs" : [
                    0.2,
                    0.5
                  ]
                }
              ]
            }
          }
        }
      }
    }
  }
}
